test_main_objs.py

```diff
@@ -0,0 +1,228 @@
+    'https://crimproject.org/mei/CRIM_Mass_0020_4.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0020_5.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0021_1.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0021_2.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0021_3.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0021_4.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0021_5.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0022_1.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0022_2.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0022_3.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0022_4.mei',
+    'https://crimproject.org/mei/CRIM_Mass_0022_5.mei'
+]
+
+def build_crim_models(files):
+    """
+    Build a list of crim models for the selected files.
+    """
+    root = "https://raw.githubusercontent.com/CRIM-Project/CRIM-online/master/crim/static/mei/MEI_3.0/"
+    prefix = "CRIM_Model_00"
+    postfix = ".mei"
+    models = {}
+    for file in files:
+        try:
+            if not "http" in file:
+                corpus = CorpusBase([root + prefix + file + postfix])
+            else:
+                file = file.split("/")[-1]
+                corpus = CorpusBase([root + file])
+            model = corpus.scores[0]
+            models[file] = model
+        except:
+            print("Can't import file", file)
+
+    return models
+
+def validate_ngrams_last_offsets(model, df, n, how='columnwise', other=None, held='Held',
+                                 exclude=['Rest'], interval_settings=('d', True, True), unit=0):
+
+    df1 = model.getNgrams(df=df, n=n, how=how, other=other, held=held,
+                  exclude=exclude, interval_settings=interval_settings, unit=unit,
+                  offsets='first')
+
+    df2 = model.getNgrams(df=df, n=n, how=how, other=other, held=held,
+                  exclude=exclude, interval_settings=interval_settings, unit=unit,
+                  offsets='last')
+
+    # compare patterns
+    df1_cols = [df1.iloc[:, i] for i in range(len(df1.columns))]
+    df2_cols = [df1.iloc[:, i] for i in range(len(df2.columns))]
+
+    # making sure that we would get the same pattern whichever offsets we choose.
+    for i in range(len(df1.columns)):
+        df1_cols[i].dropna(inplace=True)
+        df2_cols[i].dropna(inplace=True)
+
+        df1_cols[i].reset_index(drop=True)
+        df2_cols[i].reset_index(drop=True)
+
+        assert(df1_cols[i].equals(df2_cols[i]))
+
+def has_diff_measure_offsets(model):
+    # get columns
+    measures = model.getMeasure()
+    # compare each columns, return FALSE for strange columns
+    voices = measures.columns
+    first_voice = measures[voices[0]]
+
+    for voice in voices[1:]:
+        if not first_voice.equals(measures[voice]):
+            return True
+    return False
+
+def test_main():
+    files = FILES_FEW
+    models = build_crim_models(files)
+
+    for file, model in models.items():
```

Comment:
This is a good start, but it would be nice to make sure that some of the values are the same too. Maybe you could store the first two rows, two middle rows, and the last two rows of a couple of dataframes (i.e. hardcode them as part of the test suite), then rerun those .get methods on a couple of pieces, select the same rows that you hard coded, and assert that they're equal.
The tests you have are useful, but the last three lines in this function are just telling us that ts, ms, and sc execute without raising an error.

Reply:
That's a good idea. I've added the objective of the tests, and then a file that contains the hardcoded test case to not clutter the main file.